### PR TITLE
Generalize Http2Cat

### PR DESCRIPTION
**Pull request by Tratcher**
#14894 This is going to be fun.

We want to re-use Kestrel's low level Http2 test infrastructure to test HttpSys and other servers. Stephen already got it partially separated into a sample project, but that relied heavily on InternalsVisibleTo. This change takes the next step change from InternalsVisibleTo to shared src files. Moving content to the repo Shared folder gives us a clear separation of what's shared and why.

There are now three types of shared src:
- Http2: Production code shared with corefx, Http2Cat, Kestrel, and HttpSys tests.
- ServerInfrastructure: Production code shared with Http2Cat, Kestrel, and HttpSys tests.
- Http2Cat: Test client specific code used by the Http2Cat sample and HttpSys tests.

There is one API that did need to become public: The client SocketConnectionFactory.

I've included two new HttpSys tests for the 3.1 GoAway feature. I even found a bug in the process #17420.

I did add some more OS version test infrastructure (https://github.com/aspnet/Extensions/pull/2715) that needs a dependency update, but that can happen later because the affected test is already skipped due to a product bug.

### CODE REVIEW

File: src/Servers/HttpSys/test/FunctionalTests/Http2Tests.cs
Comment (jkotalik) [inline src/Servers/HttpSys/test/FunctionalTests/Http2Tests.cs]: So is there a goaway sent at all right now?

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (jkotalik) [inline src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: I'm fine with making this public in 5.0 for now, but is there any other option besides this? Does InternalsVisibleTo not work here?

File: src/Shared/Http2cat/Http2CatHostedService.cs
Comment (jkotalik) [inline src/Shared/Http2cat/Http2CatHostedService.cs]: Nice!

File: src/Servers/HttpSys/test/FunctionalTests/Http2Tests.cs
Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/Http2Tests.cs]: I followed up with them on e-mail and I'll update the test. Http.Sys sends the initial GoAway, but considers the final GoAway redundant since the client already sent one. They will only send a final GoAway if the client sends 100 streams after the initial GoAway.

File: src/Shared/Http2cat/Http2CatReadMe.md
Comment (jkotalik) [inline src/Shared/Http2cat/Http2CatReadMe.md]:
```suggestion
## Http2Cat
```

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: InternalsVisibleTo isn't really an option when we need to consume this across several projects, it doesn't scale. Stephen said they wanted to make this public soon anyways.

File: src/Shared/Http2cat/Http2Utilities.cs
Comment (jkotalik) [inline src/Shared/Http2cat/Http2Utilities.cs]: Sucks that we have so much duplicated code between this and Http2TestBase: https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TestBase.cs

File: src/Shared/Http2cat/Http2ManualFrameWriter.cs
Comment (jkotalik) [inline src/Shared/Http2cat/Http2ManualFrameWriter.cs]: I _really_ don't like that this isn't shared with the Http2FrameWriter, considering around 90% of it is the same.

Comment (jkotalik) [reply src/Shared/Http2cat/Http2ManualFrameWriter.cs]: Can we do partials to add in functionality necessary in Kestrel?

File: src/Shared/Http2cat/SslDuplexPipe.cs
Comment (jkotalik) [inline src/Shared/Http2cat/SslDuplexPipe.cs]: Why was it necessary to duplicate this? I think we need to figure out a better story for all of our DuplexPipes (put all of them in a shared code place).

Comment (jkotalik) [reply src/Shared/Http2cat/SslDuplexPipe.cs]: Nvrm, I see you moved a lot to ServerInfrastructure folder. Can this live there as well?

**Review by jkotalik — CHANGES_REQUESTED** [review]
The lack of FrameWriter sharing is my biggest concern with this PR. Other than that, the separation seem correct.

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: I know we wanted to improve this API to take in a propertybag/featurecollection. That may be something we want to consider before making this public?

File: src/Shared/Http2cat/Http2Utilities.cs
Comment (Tratcher) [reply src/Shared/Http2cat/Http2Utilities.cs]: It might be possible to re-unify them, but we'd have to work out how to toggle between sockets and in-memory. This PR is already big enough, we'll follow up on that later.

File: src/Shared/Http2cat/Http2ManualFrameWriter.cs
Comment (Tratcher) [reply src/Shared/Http2cat/Http2ManualFrameWriter.cs]: This is where I had to draw the line. The Frame types and parsers didn't have any dependencies, but Http2FrameWriter pulls in the entire Kestrel object graph of things we didn't want/need (flow control, streams, header structures, etc.. You're right, removing those dependencies removed about 10-20% of the code here.

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: We'll run it by API review Monday and see. If it's just additions then there's no need to block this PR.

File: src/Shared/Http2cat/Http2ManualFrameWriter.cs
Comment (Tratcher) [reply src/Shared/Http2cat/Http2ManualFrameWriter.cs]: Removed. It turns out I only needed two of the static APIs.

File: src/Shared/Http2cat/SslDuplexPipe.cs
Comment (Tratcher) [reply src/Shared/Http2cat/SslDuplexPipe.cs]: Moved.

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: Scratch that, Servers aren't subject to API review yet. I'll check the current API with @halter73 and file any follow up issues. We have the entire release to work out the API surface and I want to get this in to unblock HttpSys feature development.

**Review by jkotalik — APPROVED** [review]
Much better

File: src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs
Comment (anurse) [reply src/Servers/Kestrel/Transport.Sockets/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.netcoreapp.cs]: Given the namespace soup here (that it's deep within `.Kestrel.Transport.Sockets.Client`) I feel OK making this public since it won't collide with a more "real" client factory.

**Pull request merged**
